File: benchmarks/live_suite/qsort.py
import sys
sys.setrecursionlimit(200000)

def qsort_custom(arr, low, high):
    if low < high:
        pivot = arr[high]
        i = low
        for j in range(low, high):
            if arr[j] <= pivot:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
        arr[i], arr[high] = arr[high], arr[i]
        pi = i
        if pi > 0:
            qsort_custom(arr, low, pi - 1)
        qsort_custom(arr, pi + 1, high)

if __name__ == "__main__":
    n = 20000 # 20k for python to avoid excessive run times
    arr = [((i * 1664525 + 1013904223) % 1000000) for i in range(n)]
    qsort_custom(arr, 0, n - 1)
    print(f"Qsort Checksum: {arr[n // 2]}")

File: benchmarks/live_suite/qsort.rs
fn qsort_custom(arr: &mut [i64], low: isize, high: isize) {
    if low < high {
        let pivot = arr[high as usize];
        let mut i = low;
        for j in low..high {
            if arr[j as usize] <= pivot {
                arr.swap(i as usize, j as usize);
                i += 1;
            }
        }
        arr.swap(i as usize, high as usize);
        let pi = i;
        if pi > 0 {
            qsort_custom(arr, low, pi - 1);
        }
        qsort_custom(arr, pi + 1, high);
    }
}

fn main() {
    let n = 100000;
    let mut arr = Vec::with_capacity(n);
    for i in 0..n {
        arr.push(((i as i64 * 1664525 + 1013904223) % 1000000) as i64);
    }
    qsort_custom(&mut arr, 0, (n - 1) as isize);
    println!("Qsort Checksum: {}", arr[n / 2]);
}
